### ISSUE TITLE
ShardedDDP: automatic gradient reduce 

### ISSUE THREAD

**Issue by blefaudeux** [facebookresearch/fairscale#160]
## 🚀 Feature
Change ShardedDDP so that the gradient reduction happens automatically during the BW pass

## Motivation
Changing DDP is already demanding for most frameworks using fairscale, changing the training loop on top of that to add a manual reduce makes it even more so. 

## Pitch
Use either autograd functions or backward_hook registration to reduce the grads properly and release the buffers after the reduction is done

## Alternatives
Current ShardedDDP is less likely to be used

**Comment by blefaudeux** [comment]
blocked by #132 , not specific to this branch but I think that this needs to be fixed as part of this PR

**Comment by myleott** [comment]
In case it's helpful, you can use `Variable._execution_engine.queue_callback` for this purpose. This was how it was implemented a long time ago in PyTorch before they moved everything to C++, and also how it was implemented in fairseq until recently: https://github.com/pytorch/fairseq/blob/54baa2e72d1228ca7d40b78f676ed782285ae6c7/fairseq/legacy_distributed_data_parallel.py#L172-L180

The mechanics are a bit clunky -- you register a backward hook on every parameter, since you don't know which parameters will be used in the forward pass -- and each backward hook queues the callback. Then the callback has a gate to ensure it's only executed once.

**Comment by myleott** [comment]
Ah, I hadn't seen #157.

Some downsides of Functions (in general) are that they don't support kwarg inputs and only allow returning a single Tensor or a tuple of Tensors. For example, a model that returns `(x, y, None)` wouldn't work because `None` isn't a Tensor. A model that returns a dictionary would similarly not work, so this adds a lot of extra constraints on the model. The kwarg input and mixed Tuple issues can be worked around (I have a diff that fixes this for activation checkpointing [here (FB-internal link, sorry)](https://github.com/fairinternal/fairseq-py/pull/1378)), but it's pretty hacky.

The queue_callback function is undocumented and I don't have many details about it. It was previously used in PyTorch for this exact purpose, and we never had issues using it in fairseq. I agree it adds a lot of unnecessary hooks -- perhaps there's a way to simply this by adding hooks on the output Tensors instead of every module parameter... would need to think about it more.

**Comment by blefaudeux** [comment]
Thanks for all the context, very much appreciated ! Speed-wise the Functions option seem ok on a single node, but too slow when multiple nodes are involved, as per some tests on FB infra, probably because it looses some bw/coms overlap. The fact that the inputs/outputs are restricted to Tensors has also already proven to be an issue, so more work to do and your pointers will help

**Comment by blefaudeux** [comment]
@myleott spending some time on the legacy fairseq implementation that you linked, it reduces all the gradients in one go, right ? I'm mostly interested in reducing when walking back the BW pass, early tests with the linked PR (which also reduces in one batch) showed that while it works, it's too slow when inter-node communications are involved. I'm testing a per-parameter reduce hook [here](https://github.com/facebookresearch/fairscale/blob/oss_autograd_bw_overlap/fairscale/nn/data_parallel/sharded_ddp.py#L100) to try to overlap a bit better the reduce and BW computation and am curious of your thoughts (disclaimer: very early and seems wrong, debugging as I write)
